File: backend/app/api/routers/regime.py
from __future__ import annotations

from fastapi import APIRouter, Depends
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.security import require_api_key
from app.db.base import get_db
from app.db.models import RegimeSnapshot
from app.market_data.instruments import DEFAULT_INSTRUMENTS

router = APIRouter(prefix="/api/regime", tags=["regime"], dependencies=[Depends(require_api_key)])


@router.get("/current")
async def current_regime(session: AsyncSession = Depends(get_db)):
    out = {}
    for spec in DEFAULT_INSTRUMENTS:
        row = (
            await session.execute(
                select(RegimeSnapshot).where(RegimeSnapshot.symbol == spec.symbol).order_by(RegimeSnapshot.time.desc()).limit(1)
            )
        ).scalar_one_or_none()
        if row is not None:
            out[spec.symbol] = {
                "time": row.time,
                "trend_label": row.trend_label,
                "vol_label": row.vol_label,
                "confidence": row.confidence,
                "features": row.features,
            }
    return out

File: backend/src/api/routes/regime.ts
import type { FastifyInstance } from "fastify";
import { requireApiKey } from "../../core/security.js";
import { DEFAULT_INSTRUMENTS } from "../../marketData/instruments.js";
import { getLatestRegimeSnapshot } from "../../engine/regimeSnapshotCache.js";

/** Shared by GET /api/regime/current and the assistant's get_regime_current tool. */
export function getCurrentRegime(): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const spec of DEFAULT_INSTRUMENTS) {
    const row = getLatestRegimeSnapshot(spec.symbol);
    if (row) {
      out[spec.symbol] = { time: row.time, trendLabel: row.trendLabel, volLabel: row.volLabel, confidence: row.confidence, features: row.features };
    }
  }
  return out;
}

export async function regimeRoutes(app: FastifyInstance): Promise<void> {
  app.addHook("preHandler", requireApiKey);

  app.get("/api/regime/current", async () => getCurrentRegime());
}
